### PR TITLE
WIP: CI: switch Emscripten job to Meson

### PR DESCRIPTION
**Pull request by rgommers**
This is one of the last CI jobs to be converted to Meson. Still a work in progress.

Note that we need to cross-compile here. Using `pyodide build` fails on not having a cross file. The `source()` function in `pyodide_build/cli/build.py` needs a `--backend-flags` parameter so that we can pass config-settings arguments to the backend. In the meantime, we use `python -m build ...`.

One can also try this locally with:
```
$ python -m build -wnx -Csetup-args=--cross-file=\$PWD/tools/ci/emscripten.meson.cross  -Csetup-args=-Dallow-noblas=true
```
as long as both `pyodide-build` and the exact version of Emscripten (currently 3.1.32, can be installed with https://emscripten.org/docs/tools_reference/emsdk.html) are available. The configure stage of the build looks fine, the build itself fails halfway through due to:
```
include/python3.11/pyport.h:601:2: error: "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."
#error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."
```
That issue is due to a static assert in `pyconfig.h`, also discussed at: https://github.com/pyodide/pyodide/pull/2494#issuecomment-1118500743

See also https://github.com/pyodide/pyodide/pull/2238 and the patches to `_numpyconfig.h` in https://github.com/pyodide/pyodide/tree/main/packages/numpy. 

tl;dr this needs some work, and it seems like cross-compilation is hard because Pyodide doesn't ship its own Python interpreter that we can actually target during a direct cross build.

Cc @hoodmane @rth. I hope I can bother you with this. If I got any of the above wrong or if you've got a suggestion to get past the `LONG_BIT` error, I'd love to hear it. Also, a couple of questions:
- If I didn't overlook a way to pass `--config-settings` flags to the backend in the `pyodide build` CLI, would you accept a PR to Pyodide to add that? It should match [`backend-flags` in `meta.yaml` recipes](https://github.com/pyodide/pyodide/blob/8b7a9d5e3796d7abc4317f1fc21c6ffbe6c6bf1a/packages/scikit-image/meta.yaml#L17) I think.
- Since scikit-image can be cross-compiled, I had expected to be able to make that work for `numpy` too. But after reading [this comment](https://github.com/pyodide/pyodide/pull/2494#issuecomment-1118500743), that may not be true? Is that still the case, and do you see a road to a regular cross-compile working for NumPy? It's actually not clear to me how the `distutils`-based build works right now, it doesn't seem to be doing the "playback" thing explained in that comment.

### CODE REVIEW

File: tools/ci/emscripten.meson.cross [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+sdl2-config = ['emconfigure', 'sdl2-config']
+
+[host_machine]
+system = 'emscripten'
+cpu_family = 'wasm32'
+cpu = 'wasm'
+endian = 'little'
+
+[properties]
+longdouble_format = 'IEEE_QUAD_LE'
```
Comment (rgommers) [inline tools/ci/emscripten.meson.cross:17]: This is the only addition compared to the upstream `emscripten.meson.cross` file in Pyodide - necessary for now.

**Comment by hoodmane** [comment]
cc @ryanking13 @henryiii

**Comment by rgommers** [comment]
I decided that it's easier to finish this only after upgrading the build of numpy that Pyodide does in-tree to Meson and 1.26.x first. I've made a start on that, but I need to do some more digging. There was also a `pip` bug that made that a little more difficult. And https://github.com/pyodide/pyodide/issues/4118 is the same `LONG_BIT` error as I'm seeing in this PR.

Next steps I plan to take (not necessarily in this order):
- disable this CI job and remove all `setup.py` files
- finish unvendoring `meson-python` from this repo after the next meson-python release
- upgrade `numpy` in Pyodide to 1.26.0 + Meson
- if needed, ensure the out-of-tree Pyodide build can take the same `backend-flags` as the in-tree build
- re-enable this CI job on the `main` branch

**Comment by rth** [comment]
Thanks for working on this @rgommers ! I don't have useful suggestions about the errors you are facing but I agree it would probably be easier to do this in the in-tree build first. Also cc @lesteve who worked on the scikit-image meson build.

> If I didn't overlook a way to pass --config-settings flags to the backend in the pyodide build CLI, would you accept a PR to Pyodide to add that?

I think overall we could be happy to accept any changes needed to make numpy with meson work. But since `pyodide build` aims to be a drop-in replacement for `python -m build`, we need to double check in particular whether this would still be OK with cibuildwheel, or if there is some other recommended way of passing those config settings.

**Comment by rgommers** [comment]
> But since `pyodide build` aims to be a drop-in replacement for `python -m build`, we need to double check in particular whether this would still be OK with cibuildwheel, or if there is some other recommended way of passing those config settings.

That should be fine for `cibuildwheel` I think. `python -m build -Ckey=value` will add `key=value` to the `config_settings` dict and then pass that on to the backend. `-C`/`--config-setting` is implemented in `build`, and `-C`/`--config-settings` in `pip` (note the extra `s` for `pip` in the long form). I am missing that flag from `pyodide build --help`.

**Comment by rgommers** [comment]
Thank you @ryanking13, that's exactly what I needed.

**Comment by hoodmane** [comment]
@ryanking13 Do you know what the status of this is? It would be good if we could look into this again. I would look into it myself but so far I don't know much about meson...

**Comment by rgommers** [comment]
I've been meaning to get back to this PR. I believe with @ryanking13's changes for improved Meson/meson-python support in Pyodide, this should be unblocked. Timing wise I realistically won't revisit it before the NumPy 2.0 release settles down though (any week now ...).

**Comment by ryanking13** [comment]
Yes, now that we're building numpy with meson in-tree, I don't think it's going to be hard to make it work in out-of-tree too. We've recently added a patch for python-meson >= 0.15 so I think we need to release pyodide-build (0.25.1).

> before the NumPy 2.0 release settles down

Maybe emscripten CI should also wait for numpy 2.0 to be released? I've seen Pyodide builds break in many cases when packages do major releases, so it might be better to wait and see if the numpy 2.0 build works in-tree before applying it.

**Comment by rgommers** [comment]
> Maybe emscripten CI should also wait for numpy 2.0 to be released? I've seen Pyodide builds break in many cases when packages do major releases, so it might be better to wait and see if the numpy 2.0 build works in-tree before applying it.

That seems reasonable to me.

File: .github/workflows/emscripten.yml [hunk 2]
```diff
@@ -52,14 +52,23 @@ jobs:
           actions-cache-folder: emsdk-cache
 
       - name: Install pyodide-build
-        run: pip install "pydantic<2" pyodide-build==$PYODIDE_VERSION
+        run: pip install "pydantic<2" build pyodide-build==$PYODIDE_VERSION
```
Comment (rgommers) [inline .github/workflows/emscripten.yml:55]: `build` isn't needed anymore, I think `pyodide-build` will do fine.

File: .github/workflows/emscripten.yml [hunk 3]
```diff
@@ -52,14 +52,23 @@ jobs:
...
 
       - name: Build
         run: |
-          # Pyodide is still in the process of adding better/easier support for
-          # non-setup.py based builds.
-          cp pyproject.toml.setuppy pyproject.toml
-          CFLAGS=-g2 LDFLAGS=-g2 pyodide build
+          # Note that we need to cross-compile here. Using `pyodide build`
+          # fails on not having a cross file. The `source() function in
```
Comment (rgommers) [inline .github/workflows/emscripten.yml:60]: This is no longer true, the cross file is shipped by pyodide now.

File: .github/workflows/emscripten.yml [hunk 4]
```diff
@@ -52,14 +52,23 @@ jobs:
...
+          # that we can pass config-settings arguments to the backend.
+          # We'll then run into static assert issues in `pyconfig.h` due to:
+          # https://github.com/pyodide/pyodide/pull/2494#issuecomment-1118500743
+          # See also https://github.com/pyodide/pyodide/pull/2238 and the patches
+          # to _numpyconfig.h in https://github.com/pyodide/pyodide/tree/main/packages/numpy
+          #
+          # tl;dr this needs some work, and it seems like cross-compilation is
+          # hard because Pyodide doesn't ship its own Python interpreter that
+          # we can actually target during a direct cross build.
+          python -m build --wheel -Csetup-args=--cross-file=$PWD/tools/ci/emscripten.meson.cross -Csetup-args=-Dallow-noblas=true
```
Comment (rgommers) [inline .github/workflows/emscripten.yml:71]: The `-Dallow-noblas=true` is no longer needed, the fallback to internal code when BLAS is missing is automatic.

Comment (rgommers) [reply .github/workflows/emscripten.yml:71]: The `-Csetup-args=--cross-file=$PWD/tools/ci/emscripten.meson.cross` will still be needed I think, only to add the longdouble format, just in slightly different format. From [this comment](https://github.com/numpy/numpy/pull/24603#issuecomment-1713793418) I think it should be:
```
pyodide build setup-args="--cross-file=$PWD/tools/ci/emscripten.meson.cross"
```

File: .github/workflows/emscripten.yml [hunk 5]
```diff
@@ -77,4 +86,4 @@ jobs:
         run: |
           source .venv-pyodide/bin/activate
           cd ..
-          python numpy/runtests.py -n -vv
+          pytest --pyargs numpy -m "not slow"
```
Comment (rgommers) [inline .github/workflows/emscripten.yml:89]: This change is still needed.

File: meson_cpu/meson.build [hunk 6]
```diff
@@ -106,6 +107,7 @@ max_features_dict = {
   's390x': S390X_FEATURES,
   'arm': ARM_FEATURES,
   'aarch64': ARM_FEATURES,
+  'wasm32': {},
```
Comment (rgommers) [inline meson_cpu/meson.build:110]: the changes in this file will still be needed.

File: tools/ci/emscripten.meson.cross (hunk 1 above)
Comment (rgommers) [reply tools/ci/emscripten.meson.cross:17]: Most of this file can be removed, only this is needed:
```ini
[properties]
longdouble_format = 'IEEE_QUAD_LE'
```

**Review by rgommers — COMMENTED** [review]
I added some comments to this file about what is still relevant. @agriyakhetarpal is going to have a fresh attempt at a new/updated CI job.

File: .github/workflows/emscripten.yml (hunk 3 above)
Comment (rgommers) [reply .github/workflows/emscripten.yml:60]: Update: still true actually, the cross file won't get applied as long as we have `vendored-meson`.

We also need the fix from https://github.com/pyodide/pyodide/pull/4502 to actually pick up `vendored-meson`. With that, an out of tree build should work.

Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:60]: I am on it – I shall be adding a patch file to port the above PR into `vendored-meson`, as discussed on Slack.

File: .github/workflows/emscripten.yml (hunk 4 above)
Comment (ryanking13) [reply .github/workflows/emscripten.yml:71]: Pyodide bundles emscripten.meson.cross used in in-tree build, so probably you can try that one too.

```
PYODIDE_MESON_CROSS_FILE=`pyodide config get meson_cross_file`
pyodide build -C "setup-args=\"${PYODIDE_MESON_CROSS_FILE}\""
```

**Comment by hoodmane** [comment]
Thanks for working on this @rgommers  and @agriyakhetarpal!
